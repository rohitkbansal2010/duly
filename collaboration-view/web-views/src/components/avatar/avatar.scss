@import "~@styles/colors.scss";
@import "~@styles/typography.scss";

.avatar {
  display: flex !important;
  flex-shrink: 0 !important;
  border-radius: 50%;

  img {
    z-index: 2;
    width: 100%;
    height: 100%;
  }

  &Fallback {
    position: absolute;
    top: 0;
    right: 0;
    bottom: 0;
    left: 0;
    z-index: 1 !important;
  }

  &Lettered {
    position: absolute;
    top: 0;
    right: 0;
    bottom: 0;
    left: 0;
    display: flex !important;
    align-items: center !important;
    justify-content: center !important;
    width: 100% !important;
    height: 100% !important;
    color: $white;
    font-weight: 600;
    line-height: 100%;
    border-radius: 50%;
    @include nexa-extra-bold-highlight;

    &MedicalDoctor,
    &RegularEmployee {
      margin: 0 auto !important;
    }

    &MedicalDoctor {
      background-color: $dupage-dark-blue;
    }

    &RegularEmployee {
      background-color: $dupage-magenta;
    }

    &Transparent {
      background-color: transparent;
    }
  }
}

.borderAvatar {
  border-color: $white;
  border-style: solid inset solid solid;

  &RegularEmployee,
  &MedicalDoctor,
  &NonEmployee {
    @extend .borderAvatar;
    outline-style: solid;
  }

  &NonEmployee {
    outline-color: $white;
    box-shadow: 0 0.125rem 0.25rem -0.0625rem rgb(102 110 125 / 6%),
      0 0.25rem 0.375rem -0.0625rem rgb(102 110 125 / 10%), 0 0 0.0625rem 0 rgb(102 110 125 / 6%);
  }

  &RegularEmployee {
    outline-color: $dupage-violet;
  }

  &MedicalDoctor {
    outline-color: $dupage-light-blue;
  }
}
