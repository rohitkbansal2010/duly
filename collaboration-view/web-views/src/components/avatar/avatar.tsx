import { useState } from 'react';
import { Image } from 'react-bootstrap';

import { AvatarByRole } from '@enums';
import { nonEmployeeIcon } from '@icons';

import styles from './avatar.scss';

type AvatarProps = {
  src?: string;
  alt?: string;
  width?: number;
  role?: AvatarByRole;
  hasBorder?: boolean;
}

const INITIALS_LENGTH = 2;

const getInitials = (nameString: string) =>
  nameString
    .split(' ')
    .slice(0, INITIALS_LENGTH)
    .map(name =>
      name[0])
    .join('')
    .toUpperCase();

export const Avatar = ({
  src,
  alt = 'avatar',
  width = 1.5,
  role = AvatarByRole.REGULAR_EMPLOYEE,
  hasBorder = false,
}: AvatarProps) => {
  const [ error, setError ] = useState<boolean>(false);
  const [ loaded, setLoaded ] = useState<boolean>(false);
  const isImageShown = src && !error;

  const avatarClassName = []
    .concat(hasBorder ? styles[`borderAvatar${role}`] : '')
    .concat(styles.avatar)
    .join(' ');

  const letteredClassName = []
    .concat(styles.avatarLettered)
    .concat(loaded ? styles.avatarLetteredTransparent : styles[`avatarLettered${role}`])
    .join(' ');

  return (
    <div
      className={avatarClassName}
      style={{
        width: `${width}rem`,
        height: `${width}rem`,
        position: 'relative',
        fontSize: `${width / 3}rem`,
        borderWidth: `${width * 0.02}rem`,
        outlineWidth: `${width * 0.03}rem`,
      }}
    >
      {isImageShown && (
        <Image
          src={src}
          alt={alt}
          title={alt}
          width="100%"
          height="100%"
          roundedCircle
          onLoad={() =>
            setLoaded(true)}
          onError={() =>
            setError(true)}
        />
      )}
      {(!loaded || error) && (
        role === AvatarByRole.NON_EMPLOYEE
          ? (
            <Image
              className={styles.avatarFallback}
              src={nonEmployeeIcon}
              alt={alt}
              title={alt}
              roundedCircle
            />
          )
          : (
            <div className={letteredClassName}>
              {!isImageShown && getInitials(alt)}
            </div>
          )
      )}
    </div>
  );
};
