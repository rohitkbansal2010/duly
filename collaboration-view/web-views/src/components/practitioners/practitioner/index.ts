export { Practitioner } from './practitioner';
export type { } from './practitioner';
