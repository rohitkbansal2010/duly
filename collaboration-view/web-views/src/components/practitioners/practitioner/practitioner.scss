@import "~@styles/colors.scss";
@import "~@styles/main.scss";

.practitioner {
  display: flex;
  flex-direction: column;
  align-items: center;
  margin: 0 0 1.6rem;
  background: none;
  border: none;

  &Horizontal {
    width: 4rem;
    margin: 0 2.25rem;
  }

  &Other {
    flex-direction: row !important;
    width: unset;
    margin: 0 0 1.6rem;

    &HumanName {
      margin-left: 1rem;
    }
  }

  &Info {
    display: flex;
    flex-direction: column;
    align-items: center;
    width: max-content;

    &HumanName {
      color: $dupage-violet;
      font-size: var(--font-size-12px-4px);
      font-family: "NexaExtraBold", sans-serif;
      // line-height: 1.1rem;
    }

    &Role {
      color: $medium-gray;
      font-size: var(--font-size-12px-4px);
      font-family: "NexaRegular", sans-serif;
      // line-height: 1.1rem;
    }
  }
}
