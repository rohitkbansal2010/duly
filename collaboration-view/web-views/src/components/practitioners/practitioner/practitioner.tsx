/* eslint-disable jsx-a11y/no-noninteractive-element-interactions */
/* eslint-disable jsx-a11y/click-events-have-key-events */
import { Avatar } from '@components/avatar';
import { CareTeamMemberRole, PractitionerRole } from '@enums';
import { useLogoutModal } from '@hooks';
import { CareTeamMember, Practitioner as PractitionerType } from '@types';
import { getUserRole, getSrcAvatar, showFirstPrefixItem } from '@utils';

import styles from './practitioner.scss';

type MemberType = Omit<PractitionerType, 'role'> | Omit<CareTeamMember, 'memberType'>;

export type PractitionerPropsType = MemberType & {
  role: PractitionerRole | CareTeamMemberRole;
  isCurrentUser?: boolean;
  isTodaysAppointment?: boolean;
  isOtherPractitioner?: boolean;
};

export const Practitioner = ({
  humanName: {
    familyName,
    givenNames,
    prefixes,
  },
  photo,
  role,
  isCurrentUser = false,
  isTodaysAppointment = false,
  isOtherPractitioner = false,
}: PractitionerPropsType) =>{
  const { onShowLogoutModal } = useLogoutModal();

  const practitionerClass = []
    .concat(styles.practitioner)
    .concat(!isTodaysAppointment ? styles.practitionerHorizontal : '')
    .concat(isOtherPractitioner ? styles.practitionerOther : '')
    .join(' ');

  const practitionerHumanNameClass = []
    .concat(styles.practitionerInfoHumanName)
    .concat(isOtherPractitioner ? styles.practitionerOtherHumanName : '')
    .join(' ');

  return (
    <button
      onClick={isCurrentUser ? onShowLogoutModal : undefined}
      className={practitionerClass}
      style={{ cursor: isCurrentUser ? 'pointer' : 'default' }}
    >
      <Avatar
        src={getSrcAvatar(photo)}
        alt={`${showFirstPrefixItem(givenNames)}${familyName}`.trim()}
        width={3}
        hasBorder={isCurrentUser}
        role={getUserRole({ role })}
      />
      <div className={styles.practitionerInfo}>
        <div className={practitionerHumanNameClass}>
          {prefixes && prefixes.length
          ? `${showFirstPrefixItem(prefixes)}${familyName}`.trim()
          : givenNames.join(' ')
        }
        </div>
        <div className={styles.practitionerInfoRole}>
          {role !== PractitionerRole.UNKNOWN && role}
        </div>
      </div>
    </button>
  );
};
