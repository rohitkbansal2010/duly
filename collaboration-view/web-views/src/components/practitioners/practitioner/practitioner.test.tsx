import { shallow } from 'enzyme';
import toJson from 'enzyme-to-json';
import * as redux from 'react-redux';

import { Avatar } from '@components/avatar';
import { PractitionerRole } from '@enums';

import { Practitioner, PractitionerPropsType } from './practitioner';

const setUp = (props: PractitionerPropsType) =>
  shallow(<Practitioner {...props} />);

const mockProps: PractitionerPropsType = {
  id: 'qwerty1',
  humanName: {
    familyName: 'Morris',
    givenNames: [ 'Jane' ],
  },
  photo: {
    contentType: 'image/x-png',
    title: 'Photo',
    size: 0,
    data: 'data',
  },
  role: PractitionerRole.MA,
  isCurrentUser: false,
  isTodaysAppointment: true,
};

describe('Practitioner', () => {
  let component: ReturnType<typeof shallow>;

  beforeEach(() => {
    jest
      .spyOn(redux, 'useSelector')
      .mockReturnValue(jest.fn());
    jest
      .spyOn(redux, 'useDispatch')
      .mockReturnValue(jest.fn());

    component = setUp(mockProps);
  });

  it('should exist', () => {
    expect(Practitioner).toBeDefined();
  });

  it('should render the "Practitioner" component with role MA', () => {
    expect(toJson(component)).toMatchSnapshot();
    expect(component.find(Avatar)).toHaveLength(1);
    expect(component.text().includes(mockProps.humanName.givenNames.join(' '))).toBeTruthy();
    expect(component.text().includes(mockProps.role)).toBeTruthy();
  });
});
