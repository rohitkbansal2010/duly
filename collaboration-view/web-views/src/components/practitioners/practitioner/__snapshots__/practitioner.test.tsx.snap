// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Practitioner should render the "Practitioner" component with role MA 1`] = `
<button
  className="practitioner  "
  style={
    Object {
      "cursor": "default",
    }
  }
>
  <Avatar
    alt="Jane Morris"
    hasBorder={false}
    role="RegularEmployee"
    src="data:image/x-png;base64,data"
    width={3}
  />
  <div
    className="practitionerInfo"
  >
    <div
      className="practitionerInfoHumanName "
    >
      Jane
    </div>
    <div
      className="practitionerInfoRole"
    >
      MA
    </div>
  </div>
</button>
`;
